### PR TITLE
Fallback to displaying the TF diff when no Pulumi diff is available

### PR DESCRIPTION
**Pull request by iwahbe**
When the processes a Diff request from the Pulumi engine, it queries the Terraform provider, receiving a `shim.DiffInstance` in response. The `shim.DiffInstance` sits in the Terraform namespace. The bridge then tries to translate it back into the Pulumi namespace.

Most of the time, this is successful, and the user sees high quality details paths in their diff display. In rare cases, we fail to translate the diff and the user sees that a resource needs updating, but not which fields caused the update.

This PR adds a backstop for the above scenario:

If Pulumi is not able to specify which fields have changed but the Terraform diff does display fields that have changed (and the resource will be updated/replaced), then the bridge will log at `info` level the terraform diff and a suggestion to report the bug to the Pulumi provider or the bridge.

If the new logic is activated, that means TF specified a diff with specific fields **and** the bridge failed to translate that diff.

---

Q: Why do we need to log this info, can we not just put it in the detailed diff?

A: The detailed diff protocol only allows the provider to specify paths (in Pulumi space). Values diffs are calculated in the engine. For TF diffs where we are unable to translate to Pulumi paths, we can't use detailed diff.

### CODE REVIEW

File: pkg/tfbridge/provider.go [hunk 1]
```diff
@@ -1235,14 +1235,22 @@ func (p *Provider) Diff(ctx context.Context, req *pulumirpc.DiffRequest) (*pulum
...
-	// Ensure that outputs are deterministic to enable gRPC testing.
-	sort.Strings(replaces)
-	sort.Strings(stables)
+	sorted := func(arr []string) []string {
+		sort.Strings(arr)
+		return arr
+	}
+
+	if a, ok := diff.(interface {
+		Attributes() map[string]shim.ResourceAttrDiff
```
Comment (VenelinMartinov) [inline pkg/tfbridge/provider.go:1244]: Why is the cast here needed? `diff` is already a `shim.InstanceDiff` which should have that method

**Review by VenelinMartinov — COMMENTED** [review]
We are in the process of removing the access to the TF diff from the bridge - so this should not be possible - what is the issue you are adding this for?

**Comment by VenelinMartinov** [comment]
Perhaps instead of the TF diff we can print the State + Plan?

File: pkg/tfbridge/provider.go (hunk 1 above)
Comment (iwahbe) [reply pkg/tfbridge/provider.go:1244]: It doesn't:

https://github.com/pulumi/pulumi-terraform-bridge/blob/aa7015a81afda1fe211264f081bbdfa181f5d75b/pkg/tfshim/shim.go#L47-L53

It has `Attribute`, but we want the set here.

**Comment by t0yv0** [comment]
@iwahbe has ran this through providers ideation and we have some prior alignment here. This is indeed a very serious issue to perform update plans without explaining why the updates are happening. AS such, I support workarounds to improve the situation for the user. One notable limitation is that while ideally we would expose this additional information in the detailedDiff part of the DiffResponse, the protocol currently limits the detailedDiff parameter to only communicate indices into news and olds, not actual values; and this wouldn't work unless the dummy values were inserted into the payloads which makes the change too risky vs the workaround.

As such I'm supportive of this change but I want to make sure that it is done in good form that will facilitate removing this code later in favor of a more robust solution. TO help with that:

- where is the end-user scenario hitting this exactly, do you have a link? 
- would it be possible to reproduce this scenario in an integration level test?

Also wondering if we should scope removing this into the Accurate Previews work, where we will try to reconstruct the detaieldDIff from `(state, plan)` `cty.Value` pairs. Having an example to drill through can inform our work there.

File: pkg/tfbridge/diff.go [hunk 2]
```diff
@@ -386,3 +386,47 @@ func makeDetailedDiffExtra(
 		collectionDiffs: collectionDiffs,
 	}
 }
+
+func reportRawTFDiff(ctx context.Context, attrs map[string]shim.ResourceAttrDiff) {
```
Comment (t0yv0) [inline pkg/tfbridge/diff.go:390]: The not great part about using attrs map[string]shim.ResourceAttrDiff is that I can think of at least one false positive case (oof,I'll need to dig up that one). Where the InstanceDiff object returns change details but TF actually has a no-change plan which is the expected plan.

**Pull request closed without merging**